### DISCUSSION TITLE
Have a set of UserScripts bundled with Bromite

### DISCUSSION THREAD

**Discussion by peternrdstrm** [bromite/bromite#2070] [Ideas]
Note: I am willing to write and maintain the scripts, I want to experiment with writing them with TypeScript, for security and stability reasons.

Some of the most asked for features in Bromite are cosmetic filtering and blocking popups. Both of theese can be fixed with some simple-ish userscripts. It would be great is Bromite included some of these scripts, that would be trusted and stable.

**Comment by uazo** [comment]
I like the idea of having bundled userscripts too, but nothing can be included without a evaluation of sources, which I can't find. I too would like to try to use userscript to simplify the management of some patches, bringing them to javascript, as for instance `Multiple-fingerprinting-mitigations.patch`

> features in Bromite are cosmetic filtering and blocking popups. 

on the first point, I think it is not correct to act in javascript in android devices due to performance. on this point I am studying the code to understand how to insert the cosmetic filtering in the future, looking for a way not to be noticed by the anti-adblock scripts.

> Both of theese can be fixed with some simple-ish userscripts. 

anyway, reconnecting https://github.com/bromite/bromite/discussions/2066, if you are interested, could I propose you to try to write `Multiple-fingerprinting-mitigations.patch` in javascript?

**Reply by peternrdstrm** [reply]
> anyway, reconnecting #2066, if you are interested, could I propose you to try to write `Multiple-fingerprinting-mitigations.patch` in javascript?

Yes, I think I can try

**Reply by uazo** [reply]
> Yes, I think I can try

ok, and how would you do it, technically speaking.

**Reply by peternrdstrm** [reply]
> > Yes, I think I can try
> 
> ok, and how would you do it, technically speaking.

Well it depends. Do you want it just rewritten in javascript or expanded/changed? I think rewriting would be relatively simple, but would take a lot of time, with the research part included. I don't exactly understand your intention with the rewrite, could you elaborate on that?

**Reply by peternrdstrm** [reply]
> I like the idea of having bundled userscripts too, but nothing can be included without a evaluation of sources, which I can't find. 
>> Is this still an issue if for example I write the scripts and they are an official part of the Bromite repo?
>on the first point, I think it is not correct to act in javascript in android devices due to performance
>> I personally am more concerned with security as javascript tends to be one of the unsafeat languages to write stuff in. This is why I want to try a Typescript setup for userscript development. That might help with performance too ,altpugh I don't know by how much would that be

**Reply by uazo** [reply]
> Well it depends. Do you want it just rewritten in javascript

I try to explain:
1. for example that patch change `Element::ClientQuads` in https://github.com/bromite/bromite/blob/master/build/patches/Multiple-fingerprinting-mitigations.patch#L242 the goal is to dirty the values that javascript can obtain, so that they are as random as possible and capable of generating noise in the algorithm that any site uses to generate the browser fingerprint.

2. `Element::ClientQuads` is used here:
![image](https://user-images.githubusercontent.com/29201891/168591027-4047b478-432d-4b0f-bb4b-3514b8f8a3ed.png)
which at first glance seem like the api available from javascript.
you can get confirmation by looking in the idl files:
[third_party/blink/renderer/core/dom/element.idl](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.idl;l=118;drc=f2fbc0135f4f2954f293099378379997b474246e;bpv=1;bpt=0) 
[third_party/blink/renderer/core/dom/range.idl](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/range.idl;l=62;drc=f2fbc0135f4f2954f293099378379997b474246e;bpv=1;bpt=0?q=ange.idl&ss=chromium%2Fchromium%2Fsrc)

if all this were true, then from javascript with a userscript I could change the `getClientRects()` and `getBoundingClientRect()` function to one that works as if it were the patch (but I'm not sure yet).
potentially I could decide to also add potential known points used for fingerprinting, but simplifying (maybe) because javascript side

**Reply by peternrdstrm** [reply]
> > Well it depends. Do you want it just rewritten in javascript
> 
> I try to explain:
> 
> 1. for example that patch change `Element::ClientQuads` in https://github.com/bromite/bromite/blob/master/build/patches/Multiple-fingerprinting-mitigations.patch#L242 the goal is to dirty the values that javascript can obtain, so that they are as random as possible and capable of generating noise in the algorithm that any site uses to generate the browser fingerprint.
> 2. `Element::ClientQuads` is used here:
>    ![image](https://user-images.githubusercontent.com/29201891/168591027-4047b478-432d-4b0f-bb4b-3514b8f8a3ed.png)
>    which at first glance seem like the api available from javascript.
>    you can get confirmation by looking in the idl files:
>    [third_party/blink/renderer/core/dom/element.idl](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.idl;l=118;drc=f2fbc0135f4f2954f293099378379997b474246e;bpv=1;bpt=0)
>    [third_party/blink/renderer/core/dom/range.idl](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/range.idl;l=62;drc=f2fbc0135f4f2954f293099378379997b474246e;bpv=1;bpt=0?q=ange.idl&ss=chromium%2Fchromium%2Fsrc)
> 
> if all this were true, then from javascript with a userscript I could change the `getClientRects()` and `getBoundingClientRect()` function to one that works as if it were the patch (but I'm not sure yet). potentially I could decide to also add potential known points used for fingerprinting, but simplifying (maybe) because javascript side

I'll need to take a closer look.

**Reply by uazo** [reply]
> I'll need to take a closer look.

@AzureB1te forget this request, in the meantime I thought about how I would do it, and I came to the conclusion that any patch in javascript with userscripts would never work in workers, i.e. in contexts where scripts are not inject, so it would be wasted time as well as not possible.

indeed @csagan5 I do not understand how the various antifingerprinting extensions available can protect the browser by modifying the javascript apis, considering that the extensions do not have access to the context of workers of page (I have not found any reference in the extension documentation) and therefore it does not seem to me that they can inject javascript ...

for example [NetworkInformation](https://developer.mozilla.org/en-US/docs/Web/API/WorkerNavigator/connection) which is also available in the [context of workers](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope) how do they change it? same for [createImageBitmap](https://developer.mozilla.org/en-US/docs/Web/API/createImageBitmap).
do you know something about it?

**Reply by peternrdstrm** [reply]
> > I'll need to take a closer look.
> 
> @AzureB1te forget this request, in the meantime I thought about how I would do it, and I came to the conclusion that any patch in javascript with userscripts would never work in workers, i.e. in contexts where scripts are not inject, so it would be wasted time as well as not possible.
> 
> indeed @csagan5 I do not understand how the various antifingerprinting extensions available can protect the browser by modifying the javascript apis, considering that the extensions do not have access to the context of workers of page (I have not found any reference in the extension documentation) and therefore it does not seem to me that they can inject javascript ...
> 
> for example [NetworkInformation](https://developer.mozilla.org/en-US/docs/Web/API/WorkerNavigator/connection) which is also available in the [context of workers](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope) how do they change it? same for [createImageBitmap](https://developer.mozilla.org/en-US/docs/Web/API/createImageBitmap). do you know something about it?

After my closer look, I realized how limited my knowleadge of js apis is. For my first contribution I'll try to do something simpler, be that java or javascript

**Reply by csagan5** [reply]
> indeed @csagan5 I do not understand how the various antifingerprinting extensions available can protect the browser by modifying the javascript apis, considering that the extensions do not have access to the context of workers of page (I have not found any reference in the extension documentation) and therefore it does not seem to me that they can inject javascript ...
> 
> for example [NetworkInformation](https://developer.mozilla.org/en-US/docs/Web/API/WorkerNavigator/connection) which is also available in the [context of workers](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope) how do they change it? same for [createImageBitmap](https://developer.mozilla.org/en-US/docs/Web/API/createImageBitmap). do you know something about it?

Perhaps it's mostly snake oil / outdated mitigations? I would start from reading https://palant.info/2020/12/10/how-anti-fingerprinting-extensions-tend-to-make-fingerprinting-easier/ and maybe also checking out https://fpmon.github.io/fingerprinting-monitor/

**Reply by csagan5** [reply]
> > anyway, reconnecting #2066, if you are interested, could I propose you to try to write `Multiple-fingerprinting-mitigations.patch` in javascript?
> 
> Yes, I think I can try

I don't think that it would be a beneficial improvement for multiple reasons (performance and Javascript API openness).

**Reply by uazo** [reply]
> Perhaps it's mostly snake oil / outdated mitigations? 

I don't know, I also checked other extension sources (by the way, which one according to you is the best I check that code) but I don't find anything about the management of the worker context.

> I would start from reading https://palant.info/2020/12/10/how-anti-fingerprinting-extensions-tend-to-make-fingerprinting-easier/ and maybe also checking out https://fpmon.github.io/fingerprinting-monitor/

for the first link there are hacks that would allow me to hide from the host's javascript code, while the second, too bad there are no sources, but even there there is nothing for the workers

**Reply by uazo** [reply]
in the code of the various extensions I search, without finding, an override of [importScripts](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts), because in my opinion it is the only way to inject a script into the workers

**Reply by csagan5** [reply]
> by the way, which one according to you is the best I check that code

I cannot tell.

> in the code of the various extensions I search, without finding, an override of [importScripts](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts), because in my opinion it is the only way to inject a script into the workers

Maybe they never thought of it or there are other obstacles to prevent this from working.

**Reply by uazo** [reply]
> Maybe they never thought of it or there are other obstacles to prevent this from working.

well, for now the [specific extension](https://chrome.google.com/webstore/detail/canvas-fingerprint-defend/lanfdkkpgfjfdikkncbnojekcppdebfp) I found doesn't work in worker contexts:
![image](https://user-images.githubusercontent.com/29201891/169639676-bb14cf04-e63f-4f69-969b-eb73d3987779.png)

**Comment by csagan5** [comment]
> I am willing to write and maintain the scripts

This is a tricky proposal: it ties the security and privacy of all Bromite users to your devices not being compromised, it is a hard sell. In practice this means that nobody should put their guard down and all contributions must be kept being reviewed as anyone can have their devices compromised.

> I want to experiment with writing them with TypeScript

Transpiled to Javascript? Chromium does not run TypeScript directly?

> for security and stability reasons.

I don't think TypeScript can be more secure than Javascript.

> Some of the most asked for features in Bromite are cosmetic filtering and blocking popups.

That is not how features/bugfixes to pick are implemented though: not by (apparent) popularity but after discussing what is most important.

> It would be great is Bromite included some of these scripts, that would be trusted and stable.

It takes only 1 script to be collecting user data or doing something malicious to make the whole castle fall down. Distributing user scripts is no walk in the park.

To go back to your original question: the best candidate for migration to UserScripts would be these: https://github.com/bromite/bromite/blob/master/build/patches/Inject-scripts-for-AMP-tracking-ads-and-video.patch

The AMP script needs some work while the other one is a simple one to port to UserScript.

**Reply by peternrdstrm** [reply]
> 

Fair enough, I do understand that. 
On the Typescript security note, not, Chromium doesn't run it directly. 
It wouldn't make the scripts more secure, but as of my 
limited  understanding, it would make it less vulnerable, being a compiled and strongly typed language.

**Reply by csagan5** [reply]
Less vulnerable to developer's mistakes, not user security in any way.

**Reply by peternrdstrm** [reply]
> Less vulnerable to developer's mistakes, not user security in any way.

Well but aren't developer mistakes what cause security vulnerabilities?

**Reply by csagan5** [reply]
Chromium sandbox developer mistakes, not Javascript scripts mistakes: those run in the sandbox.

**Reply by peternrdstrm** [reply]
> Chromium sandbox developer mistakes, not Javascript scripts mistakes: those run in the sandbox.

Yeah fair, that makes sense.
